fblock.go

```diff
@@ -0,0 +1,455 @@
+	if ledger {
+		body, err = fb.ComputeLedgerBodyMR()
+	} else {
+		body, err = fb.ComputeBodyMR()
+	}
+	if err != nil {
+		return Bytes32{}, err
+	}
+
+	headerHash := sha256.Sum256(header)
+	if ledger { // Merkle leaves are flipped for the ledger keymr
+		leaves[0] = body[:]
+		leaves[1] = headerHash[:]
+	} else {
+		leaves[0] = headerHash[:]
+		leaves[1] = body[:]
+	}
+
+	tree := merkle.NewTreeWithOpts(merkle.TreeOptions{DoubleOddNodes: true, DisableHashLeaves: true})
+	if err := tree.Generate(leaves, sha256.New()); err != nil {
+		return Bytes32{}, err
+	}
+	root := tree.Root()
+	var keyMR Bytes32
+	copy(keyMR[:], root.Hash)
+	return keyMR, nil
+}
+
+// ComputeBodyMR computes the merkle root of the transactions in the body including
+// their signatures
+func (fb FBlock) ComputeBodyMR() (Bytes32, error) {
+	return fb.computeBodyMR(false)
+}
+
+// ComputeBodyMR computes the merkle root of the transactions in the body excluding
+// their signatures
+func (fb FBlock) ComputeLedgerBodyMR() (Bytes32, error) {
+	return fb.computeBodyMR(true)
+}
+
+// computeBodyMR will calculate the merkle root of all the transactions in the body.
+// If `ledger` is true, signature blocks of the transactions are excluded.
+func (fb FBlock) computeBodyMR(ledger bool) (Bytes32, error) {
+	if !fb.IsPopulated() {
+		return Bytes32{}, fmt.Errorf("not populated")
+	}
+
+	// Transactions + minute markers are included
+	leaves := make([][]byte, len(fb.Transactions)+len(fb.endOfPeriod))
+	period := 0
+	c := 0
+	for i, trans := range fb.Transactions {
+		for period < len(fb.endOfPeriod) && i != 0 && i == fb.endOfPeriod[period] {
+			period++
+			leaves[c] = []byte{FBlockMinuteMarker}
+			c++
+		}
+
+		var data []byte
+		var err error
+		if ledger { // Ledger does not marshal signature fields
+			data, err = trans.MarshalLedgerBinary()
+		} else {
+			data, err = trans.MarshalBinary()
+		}
+		if err != nil {
+			return Bytes32{}, err
+		}
+		leaves[c] = data
+		c++
+	}
+
+	for period < len(fb.endOfPeriod) {
+		period++
+		leaves[c] = []byte{FBlockMinuteMarker}
+		c++
+	}
+
+	tree := merkle.NewTreeWithOpts(merkle.TreeOptions{DoubleOddNodes: true})
+	if err := tree.Generate(leaves, sha256.New()); err != nil {
+		return Bytes32{}, err
+	}
+	root := tree.Root()
+	var bodyMR Bytes32
+	copy(bodyMR[:], root.Hash)
+	return bodyMR, nil
+}
+
+func (fb FBlock) ComputeFullHash() (Bytes32, error) {
+	data, err := fb.MarshalBinary()
+	if err != nil {
+		return Bytes32{}, err
+	}
+	return sha256.Sum256(data), nil
+}
+
+func (fb FBlock) ComputeHeaderHash() (Bytes32, error) {
+	header, err := fb.MarshalBinaryHeader()
+	if err != nil {
+		return Bytes32{}, err
+	}
+	return sha256.Sum256(header), nil
+}
+
+// Transaction efficiently finds and returns the *FactoidTransaction in fb.Transactions
+// for the given txid, if it exists. Otherwise, Transaction returns nil.
+func (fb FBlock) Transaction(txid Bytes32) *FactoidTransaction {
+	ei := sort.Search(len(fb.Transactions), func(i int) bool {
```

Comment:
Doesn't this require that fb.Transactions be sorted by ChainID? I thought these were in chronological order?

Reply:
Oh I totally messed that up. I thought I had done a sort first, so this function is incorrect.

Reply:
If looking up a specific txid isn't needed at this time I would omit this. This requires maintaining a sorted slice. and if this is needed by a client later, then they can be better off with a map going from txids to an index in the Transactions map.

Reply:
Removed